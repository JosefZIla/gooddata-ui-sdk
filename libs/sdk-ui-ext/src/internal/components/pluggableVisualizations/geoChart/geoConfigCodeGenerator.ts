// (C) 2022 GoodData Corporation
import { IInsightDefinition, insightProperties, insightVisualizationUrl } from "@gooddata/sdk-model";
import { IGeoConfig } from "@gooddata/sdk-ui-geo";
import filter from "lodash/fp/filter";
import flow from "lodash/fp/flow";
import fromPairs from "lodash/fromPairs";
import isNil from "lodash/isNil";
import toPairs from "lodash/toPairs";
import { IEmbeddingCodeContext } from "../../../interfaces/VisualizationDescriptor";
import { PropWithMeta } from "../../../utils/embeddingCodeGenerator";

const supportedGeoConfigProperties = new Set<keyof IGeoConfig>([
    "center",
    "colorMapping",
    "cooperativeGestures",
    "legend",
    "limit",
    "selectedSegmentItems",
    "separators",
    "viewport",
    "points",
    "showLabels",
]);

export function geoConfigFromInsight(insight: IInsightDefinition, ctx?: IEmbeddingCodeContext): IGeoConfig {
    const properties = insightProperties(insight);
    const controls = properties?.controls ?? {};
    const withValuesFromContext = {
        ...controls,
        ...(ctx?.settings?.separators ? { separators: ctx?.settings?.separators } : {}),
    };

    const configFromProperties = flow(
        toPairs,
        filter(([key, value]) => supportedGeoConfigProperties.has(key as any) && !isNil(value)),
        fromPairs,
    )(withValuesFromContext) as IGeoConfig;

    return {
        ...configFromProperties,
        ...mapBoxTokenPlaceholder(),
    };
}

export function mapBoxTokenPlaceholder(): IGeoConfig {
    return {
        mapboxToken: "<fill your Mapbox token here>",
    };
}

export function isGeoChart(insightDefinition: IInsightDefinition): boolean {
    const type = insightVisualizationUrl(insightDefinition).split(":")[1];

    return type === "pushpin";
}

export function geoConfigForInsightViewComponent(): PropWithMeta<IGeoConfig> | undefined {
    return {
        value: mapBoxTokenPlaceholder(),
        meta: {
            cardinality: "scalar",
            typeImport: {
                importType: "named",
                name: "IGeoConfig",
                package: "@gooddata/sdk-ui-geo",
            },
        },
    };
}
