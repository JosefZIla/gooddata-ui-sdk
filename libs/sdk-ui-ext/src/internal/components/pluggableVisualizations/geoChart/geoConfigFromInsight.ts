// (C) 2022 GoodData Corporation
import { IInsightDefinition, insightProperties } from "@gooddata/sdk-model";
import { IGeoConfig } from "@gooddata/sdk-ui-geo";
import filter from "lodash/fp/filter";
import flow from "lodash/fp/flow";
import fromPairs from "lodash/fromPairs";
import isNil from "lodash/isNil";
import toPairs from "lodash/toPairs";
import { IEmbeddingCodeContext } from "../../../interfaces/VisualizationDescriptor";

const supportedGeoConfigProperties = new Set<keyof IGeoConfig>([
    "center",
    "colorMapping",
    "cooperativeGestures",
    "legend",
    "limit",
    "selectedSegmentItems",
    "separators",
    "viewport",
    "points",
    "showLabels",
]);

export function geoConfigFromInsight(insight: IInsightDefinition, ctx?: IEmbeddingCodeContext): IGeoConfig {
    const properties = insightProperties(insight);
    const controls = properties?.controls ?? {};
    const withValuesFromContext = {
        ...controls,
        ...(ctx?.settings?.separators ? { separators: ctx?.settings?.separators } : {}),
    };

    const configFromProperties = flow(
        toPairs,
        filter(([key, value]) => supportedGeoConfigProperties.has(key as any) && !isNil(value)),
        fromPairs,
    )(withValuesFromContext) as IGeoConfig;

    return {
        ...configFromProperties,
        mapboxToken: "<fill your Mapbox token here>",
    };
}
