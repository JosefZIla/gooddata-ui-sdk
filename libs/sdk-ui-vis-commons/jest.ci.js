// (C) 2019 GoodData Corporation
/* eslint-disable @typescript-eslint/no-var-requires */
const ciBase = require("../../common/config/jest/jest.config.ci.base.js");

module.exports = {
    ...ciBase,
    testEnvironment: "jsdom",
    setupFilesAfterEnv: ["<rootDir>/jest.setup.ts"],
};
