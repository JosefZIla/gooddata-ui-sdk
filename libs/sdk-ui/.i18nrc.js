module.exports = {
    paths: ["./src/base/localization/bundles"],
    structure: true,
    intl: true,
    html: true,
    insightToReport: true,
    usage: false,
    debug: false,
    source: "src/**/*.{ts,js,tsx,jsx}",
    rules: [
        {
            pattern: [/.+/],
        },
    ],
};
