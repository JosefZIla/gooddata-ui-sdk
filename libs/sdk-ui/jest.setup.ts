// (C) 2019 GoodData Corporation
/* eslint-disable import/no-unassigned-import */
import * as raf from "raf";
import "@testing-library/jest-dom/extend-expect";

raf.polyfill();
