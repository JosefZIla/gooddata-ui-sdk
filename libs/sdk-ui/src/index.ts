// (C) 2007-2021 GoodData Corporation
/**
 * @packageDocumentation
 */
import "./polyfills";
// new exports

export * from "./base";
export * from "./execution";
export * from "./kpi";

/**
 * Common interface uses to specify number separators for the different SDK components.
 * @public
 */
export { ISeparators } from "@gooddata/sdk-backend-spi";
