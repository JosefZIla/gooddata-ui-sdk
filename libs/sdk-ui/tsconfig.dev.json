{
    "extends": "./tsconfig.build.esm.json",
    "compilerOptions": {
        "noImplicitAny": false,
        "noUnusedLocals": false,
        "noUnusedParameters": false,
        "pretty": true
    }
}
