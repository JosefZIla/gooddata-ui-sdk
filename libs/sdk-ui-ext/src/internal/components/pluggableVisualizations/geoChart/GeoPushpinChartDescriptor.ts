// (C) 2021-2022 GoodData Corporation
import { IInsightDefinition, ISettings } from "@gooddata/sdk-model";
import { BucketNames } from "@gooddata/sdk-ui";
import { IGeoPushpinChartProps } from "@gooddata/sdk-ui-geo";

import {
    IVisualizationDescriptor,
    IVisualizationSizeInfo,
    PluggableVisualizationFactory,
} from "../../../interfaces/VisualizationDescriptor";
import { IFluidLayoutDescriptor } from "../../../interfaces/LayoutDescriptor";
import { PluggableGeoPushpinChart } from "./PluggableGeoPushpinChart";
import { DASHBOARD_LAYOUT_DEFAULT_VIS_HEIGHT, MIDDLE_VISUALIZATION_HEIGHT } from "../constants";
import { BaseChartDescriptor } from "../baseChart/BaseChartDescriptor";
import {
    filtersInsightConversion,
    getInsightToPropsConverter,
    getReactEmbeddingCodeGenerator,
    insightConversion,
    singleAttributeBucketConversion,
    singleAttributeOrMeasureBucketConversion,
    sortsInsightConversion,
} from "../../../utils/embeddingCodeGenerator";
import { geoConfigFromInsight } from "./geoConfigFromInsight";
import { chartAdditionalFactories } from "../chartCodeGenUtils";

export class GeoPushpinChartDescriptor extends BaseChartDescriptor implements IVisualizationDescriptor {
    public getFactory(): PluggableVisualizationFactory {
        return (params) => new PluggableGeoPushpinChart(params);
    }

    public getSizeInfo(
        _insight: IInsightDefinition,
        layoutDescriptor: IFluidLayoutDescriptor,
        settings: ISettings,
    ): IVisualizationSizeInfo {
        return {
            width: {
                default: 6,
                min: 6,
                max: layoutDescriptor.gridColumnsCount,
            },
            height: {
                default: this.getDefaultHeight(settings.enableKDWidgetCustomHeight),
                min: this.getDefaultHeight(settings.enableKDWidgetCustomHeight),
                max: this.getMaxHeight(settings.enableKDWidgetCustomHeight),
            },
        };
    }

    public getEmbeddingCode = getReactEmbeddingCodeGenerator({
        component: {
            importType: "named",
            name: "GeoPushpinChart",
            package: "@gooddata/sdk-ui-geo",
        },
        insightToProps: getInsightToPropsConverter<IGeoPushpinChartProps>({
            location: singleAttributeBucketConversion("location", BucketNames.LOCATION),
            size: singleAttributeOrMeasureBucketConversion("size", BucketNames.SIZE),
            color: singleAttributeOrMeasureBucketConversion("color", BucketNames.COLOR),
            segmentBy: singleAttributeBucketConversion("segmentBy", BucketNames.SEGMENT),
            filters: filtersInsightConversion("filters"),
            sortBy: sortsInsightConversion("sortBy"),
            config: insightConversion(
                "config",
                {
                    typeImport: {
                        importType: "named",
                        name: "IGeoConfig",
                        package: "@gooddata/sdk-ui-geo",
                    },
                    cardinality: "scalar",
                },
                geoConfigFromInsight,
            ),
        }),
        additionalFactories: chartAdditionalFactories({
            getColorMappingPredicatePackage: "@gooddata/sdk-ui-geo",
        }),
    });

    protected getMinHeight(enableCustomHeight: boolean): number {
        if (!enableCustomHeight) {
            return DASHBOARD_LAYOUT_DEFAULT_VIS_HEIGHT;
        }
        return MIDDLE_VISUALIZATION_HEIGHT;
    }
}
