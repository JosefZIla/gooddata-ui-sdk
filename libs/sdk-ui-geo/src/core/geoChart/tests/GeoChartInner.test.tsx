// (C) 2020 GoodData Corporation
import React from "react";
import { ShallowWrapper, shallow } from "enzyme";
import { GeoChartInner, IGeoChartInnerProps, IGeoChartInnerOptions } from "../GeoChartInner";
import { RecShortcuts } from "../../../../__mocks__/recordings";
import { LegendPosition, FLUID_LEGEND_THRESHOLD, PositionType } from "@gooddata/sdk-ui-vis-commons";
import { IGeoConfig } from "../../../GeoChart";
import { DefaultColorPalette } from "@gooddata/sdk-ui";
import { getColorStrategy } from "../colorStrategy/geoChart";
import { createCategoryLegendItems } from "../GeoChartOptionsWrapper";

const { dv, geoData } = RecShortcuts.AllAndSmall;

function buildGeoChartOptions(): IGeoChartInnerOptions {
    const colorStrategy = getColorStrategy(DefaultColorPalette, [], geoData, dv);
    const categoryItems = createCategoryLegendItems(colorStrategy, "(empty)");

    return {
        geoData,
        categoryItems,
        colorStrategy,
        colorPalette: DefaultColorPalette,
    };
}

describe("GeoChartInner", () => {
    function renderComponent(
        customProps: Partial<IGeoChartInnerProps> = {},
        customConfig: Partial<IGeoConfig> = {},
    ): ShallowWrapper {
        const defaultProps: Partial<IGeoChartInnerProps> = {
            config: {
                mapboxToken: "",
                ...customConfig,
            },
            dataView: dv.dataView,
            geoChartOptions: buildGeoChartOptions(),
            height: 600,
        };
        return shallow(<GeoChartInner {...(defaultProps as any)} {...customProps} />);
    }

    it("should render GeoChartInner", async () => {
        const wrapper = renderComponent();
        expect(wrapper.find(".s-gd-geo-component").length).toBe(1);
        expect(wrapper.hasClass("flex-direction-column")).toBe(true);
    });

    it("should render GeoChartInner has flex-direction-row class", async () => {
        const props: Partial<IGeoChartInnerProps> = {
            config: {
                legend: {
                    position: "right",
                },
                mapboxToken: "",
            },
        };
        const wrapper = renderComponent(props);
        expect(wrapper.hasClass("flex-direction-row")).toBe(true);
    });

    it("should render GeoChartInner has flex-direction-column class", async () => {
        const props: Partial<IGeoChartInnerProps> = {
            config: {
                legend: {
                    position: "bottom",
                },
                mapboxToken: "",
            },
        };
        const wrapper = renderComponent(props);
        expect(wrapper.hasClass("flex-direction-column")).toBe(true);
    });

    it("should use custom Chart renderer", () => {
        const chartRenderer = jest.fn().mockReturnValue(<div />);
        renderComponent({ chartRenderer });
        expect(chartRenderer).toHaveBeenCalledTimes(1);
    });

    it("should use custom Legend renderer", () => {
        const legendRenderer = jest.fn().mockReturnValue(<div />);
        renderComponent({ legendRenderer });

        expect(legendRenderer).toHaveBeenCalledTimes(1);

        const legendProps = legendRenderer.mock.calls[0][0];
        expect(legendProps).toMatchSnapshot({
            containerId: expect.any(String),
        });
    });

    it("should return enabledLegendItems with length equal categories length", () => {
        const wrapper = renderComponent();
        expect(wrapper.state("enabledLegendItems")).toEqual([true, true, true]);
    });

    it("should call pushData", () => {
        const pushData = jest.fn();
        const props: Partial<IGeoChartInnerProps> = {
            pushData,
        };
        renderComponent(props);
        expect(pushData).toBeCalledTimes(1);
    });

    describe("GeoChart Legend", () => {
        const defaultDocumentObj: any = {
            documentElement: {
                clientWidth: FLUID_LEGEND_THRESHOLD,
            },
        };

        interface ICustomComponentProps {
            documentObj?: any;
            height?: number;
            position?: PositionType;
            responsive?: boolean | "autoPositionWithPopup";
        }

        const getCustomComponentProps = (props: ICustomComponentProps): Partial<IGeoChartInnerProps> => {
            const { height = 600, position, responsive = false, documentObj = defaultDocumentObj } = props;
            return {
                documentObj,
                config: {
                    legend: {
                        enabled: true,
                        position,
                        responsive,
                    },
                    mapboxToken: "",
                },
                height,
            };
        };

        it("should not render if legend is disabled", () => {
            const legendRenderer = jest.fn().mockReturnValue(<div />);
            renderComponent({ legendRenderer }, { legend: { enabled: false } });
            expect(legendRenderer).toHaveBeenCalledTimes(0);
        });

        it("should set flex-direction-column class for legend position TOP", () => {
            const customProps = getCustomComponentProps({ position: LegendPosition.TOP });
            const wrapper = renderComponent(customProps);
            expect(wrapper.hasClass("flex-direction-column")).toBe(true);
        });

        it("should set flex-direction-column class for legend position BOTTOM", () => {
            const customProps = getCustomComponentProps({ position: LegendPosition.BOTTOM });
            const wrapper = renderComponent(customProps);
            expect(wrapper.hasClass("flex-direction-column")).toBe(true);
        });

        it("should set flex-direction-row class for legend position LEFT", () => {
            const customProps = getCustomComponentProps({ position: LegendPosition.LEFT });
            const wrapper = renderComponent(customProps);
            expect(wrapper.hasClass("flex-direction-row")).toBe(true);
        });

        it("should set flex-direction-row class for legend position RIGHT", () => {
            const customProps = getCustomComponentProps({ position: LegendPosition.RIGHT });
            const wrapper = renderComponent(customProps);
            expect(wrapper.hasClass("flex-direction-row")).toBe(true);
        });

        it("should set responsive-legend class for responsive legend", () => {
            const customProps = getCustomComponentProps({ responsive: true });
            const wrapper = renderComponent(customProps);
            expect(wrapper.hasClass("responsive-legend")).toBe(true);
        });

        it("should set non-responsive-legend class for non responsive legend", () => {
            const customProps = getCustomComponentProps({ responsive: false });
            const wrapper = renderComponent(customProps);
            expect(wrapper.hasClass("non-responsive-legend")).toBe(true);
        });

        it("should render responsive legend for mobile", () => {
            const documentObj = {
                documentElement: {
                    clientWidth: FLUID_LEGEND_THRESHOLD - 10,
                },
            };
            const customProps = getCustomComponentProps({ responsive: true, documentObj });

            const wrapper = renderComponent(customProps);
            expect(wrapper.state("showFluidLegend")).toBeTruthy();
        });

        it("should render StaticLegend on desktop", () => {
            const documentObj = {
                documentElement: {
                    clientWidth: FLUID_LEGEND_THRESHOLD + 10,
                },
            };
            const customProps = getCustomComponentProps({ responsive: true, documentObj });

            const wrapper = renderComponent(customProps);
            expect(wrapper.state("showFluidLegend")).toBeFalsy();
        });

        it("should not set responsive-legend if responsive is autoPositionWithPopup", () => {
            const customProps = getCustomComponentProps({ responsive: "autoPositionWithPopup" });
            const wrapper = renderComponent(customProps);
            expect(wrapper.hasClass("responsive-legend")).toBe(false);
        });
    });
});
