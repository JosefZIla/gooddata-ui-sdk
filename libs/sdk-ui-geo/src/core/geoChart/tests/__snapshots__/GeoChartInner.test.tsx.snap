// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GeoChartInner should use custom Legend renderer 1`] = `
Object {
  "categoryItems": Array [
    Object {
      "color": "rgb(20,178,226)",
      "isVisible": true,
      "legendIndex": 0,
      "name": "Florida",
      "uri": "/gdc/md/liveexamplesworkspace/obj/9461/elements?id=6340105",
    },
    Object {
      "color": "rgb(0,193,141)",
      "isVisible": true,
      "legendIndex": 1,
      "name": "Texas",
      "uri": "/gdc/md/liveexamplesworkspace/obj/9461/elements?id=4436534",
    },
    Object {
      "color": "rgb(229,77,66)",
      "isVisible": true,
      "legendIndex": 2,
      "name": "California",
      "uri": "/gdc/md/liveexamplesworkspace/obj/9461/elements?id=6340116",
    },
  ],
  "colorLegendValue": "rgb(20,178,226)",
  "containerId": Any<String>,
  "contentRect": undefined,
  "format": "#,##0.00",
  "geoData": Object {
    "color": Object {
      "data": Array [
        4969,
        256,
        56,
        0,
        1856,
        1305,
        628,
        982,
        1683,
        3295,
        7286,
      ],
      "format": "#,##0.00",
      "index": 1,
      "name": "Density",
    },
    "location": Object {
      "data": Array [
        Object {
          "lat": 25.7839,
          "lng": -80.2102,
        },
        Object {
          "lat": 26.2181,
          "lng": -98.418,
        },
        Object {
          "lat": 26.2585,
          "lng": -97.695,
        },
        Object {
          "lat": 26.3024,
          "lng": -98.6425,
        },
        Object {
          "lat": 26.5281,
          "lng": -80.0811,
        },
        Object {
          "lat": 27.9942,
          "lng": -82.4451,
        },
        Object {
          "lat": 28.2405,
          "lng": -82.1796,
        },
        Object {
          "lat": 28.4772,
          "lng": -81.3369,
        },
        Object {
          "lat": 32.8312,
          "lng": -117.1225,
        },
        Object {
          "lat": 34.1139,
          "lng": -118.4068,
        },
        Object {
          "lat": 37.7562,
          "lng": -122.443,
        },
      ],
      "index": 0,
      "name": "City",
    },
    "segment": Object {
      "data": Array [
        "Florida",
        "Texas",
        "Texas",
        "Texas",
        "Florida",
        "Florida",
        "Florida",
        "Florida",
        "California",
        "California",
        "California",
      ],
      "index": 1,
      "name": "State Name",
      "uris": Array [
        "/gdc/md/liveexamplesworkspace/obj/9461/elements?id=6340105",
        "/gdc/md/liveexamplesworkspace/obj/9461/elements?id=4436534",
        "/gdc/md/liveexamplesworkspace/obj/9461/elements?id=4436534",
        "/gdc/md/liveexamplesworkspace/obj/9461/elements?id=4436534",
        "/gdc/md/liveexamplesworkspace/obj/9461/elements?id=6340105",
        "/gdc/md/liveexamplesworkspace/obj/9461/elements?id=6340105",
        "/gdc/md/liveexamplesworkspace/obj/9461/elements?id=6340105",
        "/gdc/md/liveexamplesworkspace/obj/9461/elements?id=6340105",
        "/gdc/md/liveexamplesworkspace/obj/9461/elements?id=6340116",
        "/gdc/md/liveexamplesworkspace/obj/9461/elements?id=6340116",
        "/gdc/md/liveexamplesworkspace/obj/9461/elements?id=6340116",
      ],
    },
    "size": Object {
      "data": Array [
        6381966,
        2228,
        121,
        0,
        77702,
        2804240,
        76416,
        1776841,
        3210314,
        1.28,
        3603761,
      ],
      "format": "#,##0.00",
      "index": 0,
      "name": "Population",
    },
    "tooltipText": Object {
      "data": Array [
        "Miami",
        "Abram",
        "Lasana",
        "Longoria",
        "Boynton Beach",
        "Tampa",
        "Zephyrhills",
        "Orlando",
        "San Diego",
        "Los Angeles",
        "San Francisco",
      ],
      "index": 2,
      "name": "City",
    },
  },
  "height": 600,
  "isFluidLegend": false,
  "locale": undefined,
  "maxRows": undefined,
  "name": undefined,
  "onItemClick": [Function],
  "position": "top",
  "renderPopUp": false,
  "responsive": undefined,
}
`;
